src/combinatorics/hindman.lean

```diff
@@ -0,0 +1,230 @@
+so we can repeat the argument starting from `s₁`, obtaining `a₁`, `s₂`, etc. This gives the desired
+infinite sequence. -/
+  have exists_elem : ∀ {s : set M} (hs : s ∈ U), (s ∩ { m | ∀ᶠ m' in U, m * m' ∈ s }).nonempty :=
+    λ s hs, ultrafilter.nonempty_of_mem (inter_mem hs $ by { rw ←U_idem at hs, exact hs }),
+  let elem : { s // s ∈ U } → M := λ p, (exists_elem p.property).some,
+  let succ : { s // s ∈ U } → { s // s ∈ U } := λ p, ⟨p.val ∩ { m | elem p * m ∈ p.val },
+    inter_mem p.2 $ show _, from set.inter_subset_right _ _ (exists_elem p.2).some_mem⟩,
+  use stream.corec elem succ (subtype.mk s₀ sU),
+  suffices : ∀ (a : stream M) (m ∈ FP a), ∀ p, a = stream.corec elem succ p → m ∈ p.val,
+  { intros m hm, exact this _ m hm ⟨s₀, sU⟩ rfl, },
+  clear sU s₀,
+  intros a m h,
+  induction h with b b n h ih b n h ih,
+  { rintros p rfl,
+    rw [stream.corec_eq, stream.head_cons],
+    exact set.inter_subset_left _ _ (set.nonempty.some_mem _), },
+  { rintros p rfl,
+    refine set.inter_subset_left _ _ (ih (succ p) _),
+    rw [stream.corec_eq, stream.tail_cons], },
+  { rintros p rfl,
+    have := set.inter_subset_right _ _ (ih (succ p) _),
+    { simpa only using this },
+    rw [stream.corec_eq, stream.tail_cons], },
+end
+
+/-- The strong form of **Hindman's theorem**: in any finite cover of an FP-set, one the parts
+contains an FP-set. -/
+@[to_additive FS_partition_regular]
+lemma FP_partition_regular {M} [semigroup M] (a : stream M) (s : set (set M)) (sfin : s.finite)
+  (scov : FP a ⊆ ⋃₀ s) : ∃ (c ∈ s) (b : stream M), FP b ⊆ c :=
+let ⟨U, idem, aU⟩ := exists_idempotent_ultrafilter_le_FP a in
+let ⟨c, cs, hc⟩ := (ultrafilter.finite_sUnion_mem_iff sfin).mp (mem_of_superset aU scov) in
+⟨c, cs, exists_FP_of_large U idem c hc⟩
+
+/-- The weak form of **Hindman's theorem**: in any finite cover of a nonempty semigroup, one of the
+parts contains an FP-set. -/
+@[to_additive exists_FS_of_finite_cover]
+lemma exists_FP_of_finite_cover {M} [semigroup M] [nonempty M] (s : set (set M))
+  (sfin : s.finite) (scov : ⊤ ⊆ ⋃₀ s) : ∃ (c ∈ s) (a : stream M), FP a ⊆ c :=
+let ⟨U, hU⟩ := exists_idempotent_of_compact_t2_of_continuous_mul_left
+  (@ultrafilter.continuous_mul_left M _) in
+let ⟨c, c_s, hc⟩ := (ultrafilter.finite_sUnion_mem_iff sfin).mp (mem_of_superset univ_mem scov) in
+⟨c, c_s, exists_FP_of_large U hU c hc⟩
+
+@[to_additive FS_iter_tail_sub_FS]
+lemma FP_drop_subset_FP {M} [semigroup M] (a : stream M) (n : ℕ) :
+  FP (a.drop n) ⊆ FP a :=
+begin
+  induction n with n ih, { refl },
+  rw [nat.succ_eq_one_add, ←stream.drop_drop],
+  exact trans (FP.tail _) ih,
+end
+
+@[to_additive]
+lemma FP.singleton {M} [semigroup M] (a : stream M) (i : ℕ) : a.nth i ∈ FP a :=
+by { induction i with i ih generalizing a, { apply FP.head }, { apply FP.tail, apply ih } }
+
+@[to_additive]
+lemma FP.mul_two {M} [semigroup M] (a : stream M) (i j : ℕ) (ij : i < j) :
```

Comment:
Do you need `i < j`, or is `i ≠ j` sufficient?

Reply:
You need `i < j` unless you assume `M` is commutative